File: typescript/external-session-storage/postgres-sample.ts
import { randomUUID } from "node:crypto";
import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

import {
  accessTokenFromEnv,
  deleteSession,
  getSessionInfo,
  query,
  type SDKResultSuccess,
  type SessionKey,
  type SessionStore,
  type SessionStoreEntry,
} from "@qoder-ai/qoder-agent-sdk";
import pg, { type Pool } from "pg";

const SQL_IDENTIFIER = /^[A-Za-z_][A-Za-z0-9_]*$/;

/**
 * PostgreSQL reference adapter with one row per transcript entry.
 *
 * The increasing seq column preserves the order returned by load().
 */
export class PostgresSessionStore implements SessionStore {
  constructor(
    private readonly pool: Pool,
    private readonly tableName = "qoder_session_entries",
  ) {
    if (!SQL_IDENTIFIER.test(tableName)) {
      throw new Error(
        `Invalid table name ${JSON.stringify(tableName)}. Use letters, digits, and underscores.`,
      );
    }
  }

  async ensureSchema(): Promise<void> {
    await this.pool.query(`
      CREATE TABLE IF NOT EXISTS ${this.tableName} (
        seq BIGSERIAL PRIMARY KEY,
        project_key TEXT NOT NULL,
        session_id TEXT NOT NULL,
        subpath TEXT NOT NULL DEFAULT '',
        entry JSONB NOT NULL,
        modified_at TIMESTAMPTZ NOT NULL DEFAULT clock_timestamp()
      )
    `);
    await this.pool.query(`
      CREATE INDEX IF NOT EXISTS ${this.tableName}_lookup_idx
      ON ${this.tableName} (project_key, session_id, subpath, seq)
    `);
  }

  async append(
    key: SessionKey,
    entries: SessionStoreEntry[],
  ): Promise<void> {
    if (entries.length === 0) return;

    const values: unknown[] = [
      key.projectKey,
      key.sessionId,
      key.subpath ?? "",
    ];
    const rows = entries.map((entry, index) => {
      values.push(JSON.stringify(entry));
      return `($1, $2, $3, $${index + 4}::jsonb)`;
    });
    await this.pool.query(
      `
        INSERT INTO ${this.tableName}
          (project_key, session_id, subpath, entry)
        VALUES ${rows.join(", ")}
      `,
      values,
    );
  }

  async load(key: SessionKey): Promise<SessionStoreEntry[] | null> {
    const result = await this.pool.query<{ entry: SessionStoreEntry }>(
      `
        SELECT entry
        FROM ${this.tableName}
        WHERE project_key = $1
          AND session_id = $2
          AND subpath = $3
        ORDER BY seq
      `,
      [key.projectKey, key.sessionId, key.subpath ?? ""],
    );
    return result.rows.length === 0
      ? null
      : result.rows.map(({ entry }) => entry);
  }

  async listSessions(
    projectKey: string,
  ): Promise<Array<{ sessionId: string; mtime: number }>> {
    const result = await this.pool.query<{
      session_id: string;
      modified_at: Date;
    }>(
      `
        SELECT session_id, MAX(modified_at) AS modified_at
        FROM ${this.tableName}
        WHERE project_key = $1 AND subpath = ''
        GROUP BY session_id
        ORDER BY modified_at DESC
      `,
      [projectKey],
    );
    return result.rows.map(({ session_id, modified_at }) => ({
      sessionId: session_id,
      mtime: modified_at.getTime(),
    }));
  }

  async listSubkeys(
    key: Omit<SessionKey, "subpath">,
  ): Promise<string[]> {
    const result = await this.pool.query<{ subpath: string }>(
      `
        SELECT DISTINCT subpath
        FROM ${this.tableName}
        WHERE project_key = $1
          AND session_id = $2
          AND subpath <> ''
        ORDER BY subpath
      `,
      [key.projectKey, key.sessionId],
    );
    return result.rows.map(({ subpath }) => subpath);
  }

  async delete(key: SessionKey): Promise<void> {
    if (key.subpath === undefined) {
      await this.pool.query(
        `
          DELETE FROM ${this.tableName}
          WHERE project_key = $1 AND session_id = $2
        `,
        [key.projectKey, key.sessionId],
      );
      return;
    }
    await this.pool.query(
      `
        DELETE FROM ${this.tableName}
        WHERE project_key = $1
          AND session_id = $2
          AND subpath = $3
      `,
      [key.projectKey, key.sessionId, key.subpath],
    );
  }
}

async function runQuery(options: {
  workspace: string;
  prompt: string;
  store: SessionStore;
  resume?: string;
}): Promise<SDKResultSuccess> {
  const stream = query({
    prompt: options.prompt,
    options: {
      auth: accessTokenFromEnv(),
      cwd: options.workspace,
      tools: [],
      maxTurns: 1,
      model: "auto",
      sessionStore: options.store,
      ...(options.resume ? { resume: options.resume } : {}),
    },
  });

  let result: SDKResultSuccess | undefined;
  try {
    for await (const message of stream) {
      if (message.type === "system" && message.subtype === "mirror_error") {
        throw new Error(
          `SessionStore could not persist ${JSON.stringify(message.key)}: ${message.error}`,
        );
      }
      if (message.type === "result") {
        if (message.subtype !== "success") {
          throw new Error(message.errors?.join("\n") || message.subtype);
        }
        result = message;
      }
    }
  } finally {
    await stream.close();
  }

  if (!result) throw new Error("The query ended without a success result.");
  return result;
}

export async function run(workspace: string): Promise<void> {
  const pool = new pg.Pool({
    connectionString:
      process.env["QODER_SAMPLE_POSTGRES_URL"] ??
      "postgresql://qoder:qoder@127.0.0.1:5432/qoder_session_store",
  });
  const store = new PostgresSessionStore(pool);
  const marker = `session-storage-${randomUUID()}`;
  let sessionId: string | undefined;

  try {
    await store.ensureSchema();
    console.log("[host-a] Starting a session with PostgreSQL storage.");
    const first = await runQuery({
      workspace,
      store,
      prompt: `Remember this exact deployment marker: ${marker}. Reply only that it is stored.`,
    });
    sessionId = first.session_id;
    await deleteSession(sessionId, { dir: workspace });
    console.log(
      `[host-a] Stored session ${sessionId}; local transcript deleted.`,
    );

    console.log("[host-b] Starting without host A's local transcript.");
    const resumed = await runQuery({
      workspace,
      store,
      resume: sessionId,
      prompt:
        "What exact deployment marker did I ask you to remember? Reply only with the marker.",
    });
    console.log(`[host-b] Resumed session ${resumed.session_id}.`);
    console.log(`[host-b] Agent response: ${resumed.result}`);

    if (!resumed.result.includes(marker)) {
      throw new Error("The resumed response did not contain the stored marker.");
    }
    console.log("[app] External session handoff verified.");
  } finally {
    try {
      if (sessionId) {
        const localSession = await getSessionInfo(sessionId, {
          dir: workspace,
        });
        if (localSession) {
          await deleteSession(sessionId, { dir: workspace });
        }
        await deleteSession(sessionId, {
          dir: workspace,
          sessionStore: store,
        });
        console.log("[app] Deleted sample transcript.");
      }
    } finally {
      await pool.end();
    }
  }
}

async function main(): Promise<void> {
  await run(resolve(process.argv[2] ?? process.cwd()));
}

if (
  process.argv[1] &&
  import.meta.url === pathToFileURL(process.argv[1]).href
) {
  main().catch((error: unknown) => {
    console.error(error instanceof Error ? error.message : error);
    process.exitCode = 1;
  });
}

File: python/external-session-storage/postgres_sample.py
"""Runnable PostgreSQL SessionStore reference sample."""

from __future__ import annotations

import argparse
import asyncio
import json
import os
import re
import uuid
from pathlib import Path
from typing import cast

import asyncpg
from qoder_agent_sdk import (
    QoderAgentOptions,
    ResultMessage,
    SDKMirrorErrorMessage,
    SessionKey,
    SessionListSubkeysKey,
    SessionStore,
    SessionStoreEntry,
    SessionStoreListEntry,
    access_token_from_env,
    delete_session,
    delete_session_via_store,
    get_session_info,
    query,
)

_SQL_IDENTIFIER = re.compile(r"^[A-Za-z_][A-Za-z0-9_]*$")


class PostgresSessionStore(SessionStore):
    """Store one transcript entry per PostgreSQL JSONB row."""

    def __init__(
        self,
        pool: asyncpg.Pool,
        table_name: str = "qoder_session_entries",
    ) -> None:
        if not _SQL_IDENTIFIER.fullmatch(table_name):
            raise ValueError(
                f"Invalid table name {table_name!r}. "
                "Use letters, digits, and underscores."
            )
        self._pool = pool
        self._table_name = table_name

    async def ensure_schema(self) -> None:
        """Create the entry table and lookup index when absent."""

        await self._pool.execute(
            f"""
            CREATE TABLE IF NOT EXISTS {self._table_name} (
              seq BIGSERIAL PRIMARY KEY,
              project_key TEXT NOT NULL,
              session_id TEXT NOT NULL,
              subpath TEXT NOT NULL DEFAULT '',
              entry JSONB NOT NULL,
              modified_at TIMESTAMPTZ NOT NULL DEFAULT clock_timestamp()
            );
            CREATE INDEX IF NOT EXISTS {self._table_name}_lookup_idx
              ON {self._table_name}
              (project_key, session_id, subpath, seq);
            """
        )

    async def append(
        self,
        key: SessionKey,
        entries: list[SessionStoreEntry],
    ) -> None:
        if not entries:
            return

        await self._pool.execute(
            f"""
            INSERT INTO {self._table_name}
              (project_key, session_id, subpath, entry)
            SELECT $1, $2, $3, value
            FROM unnest($4::jsonb[]) WITH ORDINALITY AS item(value, position)
            ORDER BY position
            """,
            key["project_key"],
            key["session_id"],
            key.get("subpath", ""),
            [json.dumps(entry, separators=(",", ":")) for entry in entries],
        )

    async def load(self, key: SessionKey) -> list[SessionStoreEntry] | None:
        rows = await self._pool.fetch(
            f"""
            SELECT entry
            FROM {self._table_name}
            WHERE project_key = $1
              AND session_id = $2
              AND subpath = $3
            ORDER BY seq
            """,
            key["project_key"],
            key["session_id"],
            key.get("subpath", ""),
        )
        if not rows:
            return None
        entries: list[SessionStoreEntry] = []
        for row in rows:
            value = row["entry"]
            decoded = json.loads(value) if isinstance(value, str) else value
            entries.append(cast(SessionStoreEntry, decoded))
        return entries

    async def list_sessions(
        self,
        project_key: str,
    ) -> list[SessionStoreListEntry]:
        rows = await self._pool.fetch(
            f"""
            SELECT session_id,
                   EXTRACT(EPOCH FROM MAX(modified_at)) * 1000 AS mtime
            FROM {self._table_name}
            WHERE project_key = $1 AND subpath = ''
            GROUP BY session_id
            ORDER BY mtime DESC
            """,
            project_key,
        )
        return [
            {
                "session_id": cast(str, row["session_id"]),
                "mtime": int(row["mtime"]),
            }
            for row in rows
        ]

    async def list_subkeys(self, key: SessionListSubkeysKey) -> list[str]:
        rows = await self._pool.fetch(
            f"""
            SELECT DISTINCT subpath
            FROM {self._table_name}
            WHERE project_key = $1
              AND session_id = $2
              AND subpath <> ''
            ORDER BY subpath
            """,
            key["project_key"],
            key["session_id"],
        )
        return [cast(str, row["subpath"]) for row in rows]

    async def delete(self, key: SessionKey) -> None:
        subpath = key.get("subpath")
        if subpath is None:
            await self._pool.execute(
                f"""
                DELETE FROM {self._table_name}
                WHERE project_key = $1 AND session_id = $2
                """,
                key["project_key"],
                key["session_id"],
            )
            return

        await self._pool.execute(
            f"""
            DELETE FROM {self._table_name}
            WHERE project_key = $1
              AND session_id = $2
              AND subpath = $3
            """,
            key["project_key"],
            key["session_id"],
            subpath,
        )


async def run_query(
    *,
    workspace: Path,
    prompt: str,
    store: SessionStore,
    resume: str | None = None,
) -> ResultMessage:
    options = QoderAgentOptions(
        auth=access_token_from_env(),
        cwd=workspace,
        tools=[],
        max_turns=1,
        model="auto",
        session_store=store,
        resume=resume,
    )
    result: ResultMessage | None = None
    async for message in query(prompt=prompt, options=options):
        if isinstance(message, SDKMirrorErrorMessage):
            raise RuntimeError(
                f"SessionStore could not persist {message.key}: {message.error}"
            )
        if isinstance(message, ResultMessage):
            if message.subtype != "success":
                raise RuntimeError("\n".join(message.errors or [message.subtype]))
            result = message

    if result is None:
        raise RuntimeError("The query ended without a success result.")
    return result


async def run(workspace: Path) -> None:
    pool = await asyncpg.create_pool(
        os.getenv(
            "QODER_SAMPLE_POSTGRES_URL",
            "postgresql://qoder:qoder@127.0.0.1:5432/qoder_session_store",
        )
    )
    if pool is None:
        raise RuntimeError("Could not create the PostgreSQL connection pool.")
    store = PostgresSessionStore(pool)
    marker = f"session-storage-{uuid.uuid4()}"
    session_id: str | None = None

    try:
        await store.ensure_schema()
        print("[host-a] Starting a session with PostgreSQL storage.")
        first = await run_query(
            workspace=workspace,
            store=store,
            prompt=(
                f"Remember this exact deployment marker: {marker}. "
                "Reply only that it is stored."
            ),
        )
        session_id = first.session_id
        delete_session(session_id, str(workspace))
        print(f"[host-a] Stored session {session_id}; local transcript deleted.")

        print("[host-b] Starting without host A's local transcript.")
        resumed = await run_query(
            workspace=workspace,
            store=store,
            resume=session_id,
            prompt=(
                "What exact deployment marker did I ask you to remember? "
                "Reply only with the marker."
            ),
        )
        print(f"[host-b] Resumed session {resumed.session_id}.")
        print(f"[host-b] Agent response: {resumed.result}")

        if marker not in (resumed.result or ""):
            raise RuntimeError(
                "The resumed response did not contain the stored marker."
            )
        print("[app] External session handoff verified.")
    finally:
        try:
            if session_id:
                if get_session_info(session_id, str(workspace)) is not None:
                    delete_session(session_id, str(workspace))
                await delete_session_via_store(store, session_id, workspace)
                print("[app] Deleted sample transcript.")
        finally:
            await pool.close()


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("workspace", nargs="?", type=Path, default=Path.cwd())
    workspace = cast(Path, parser.parse_args().workspace).resolve()
    asyncio.run(run(workspace))


if __name__ == "__main__":
    try:
        main()
    except (RuntimeError, OSError, ValueError) as error:
        raise SystemExit(str(error)) from error
